endpoints/event.go

```diff
@@ -0,0 +1,367 @@
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(fmt.Sprintf("invalid request: %s\n", err.Error())))
+
+		return
+	}
+
+	// validate account id
+	accountId, err := validateRequiredParameter(r, ACCOUNT_ID_PARAMETER)
+
+	if err != nil {
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(fmt.Sprintf("Account '%s' is required query parameter and can't be empty", ACCOUNT_ID_PARAMETER)))
+		return
+	}
+	eventRequest.AccountId = accountId
+
+	if eventRequest.Analytics == ENABLED {
+
+		// get account details
+		account, err := e.DataCache.Accounts().Get(eventRequest.AccountId)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				account = &cache.Account{
+					ID:            accountId,
+					EventsEnabled: false,
+				}
+			} else {
+				if glog.V(2) {
+					glog.Infof("Invalid account id: %v", err)
+				}
+
+				status := http.StatusInternalServerError
+				message := fmt.Sprintf("Invalid request: %s\n", err.Error())
+
+				w.WriteHeader(status)
+				w.Write([]byte(message))
+				return
+			}
+		}
+
+		// account does not support events
+		if !account.EventsEnabled {
+			w.WriteHeader(http.StatusUnauthorized)
+			w.Write([]byte(fmt.Sprintf("Account '%s' doesn't support events", eventRequest.AccountId)))
+			return
+		}
+
+		// handle notification event
+		e.Analytics.LogNotificationEventObject(&analytics.NotificationEvent{
+			Type:      string(eventRequest.Type),
+			Timestamp: eventRequest.Timestamp,
+			Bidid:     eventRequest.Bidid,
+			Bidder:    eventRequest.Bidder,
+			Account:   account,
+		})
+
+		// OK
+		w.WriteHeader(http.StatusOK)
+		// Add tracking pixel if format == image
+		if eventRequest.Format == IMAGE {
+			w.Header().Add("Content-Type", TRACKING_PIXEL_PNG_CONTENT_TYPE)
+			w.Write(e.TrackingPixel.Content)
+		}
+
+		return
+	}
+}
+
+/**
+ * Converts an EventRequest to an URL
+ */
+func EventRequestToUrl(externalUrl string, request *EventRequest) string {
+	s := fmt.Sprintf(TEMPLATE_URL, externalUrl, request.Type, request.Bidid, request.AccountId)
+
+	return s + optionalParameters(request)
+}
+
+/**
+ * Parses an EventRequest from an Http request
+ */
+func ParseEventRequest(r *http.Request) (*EventRequest, error) {
+	er := &EventRequest{}
+
+	// validate type
+	err := validateType(er, r)
+
+	if err != nil {
+		return er, err
+	}
+
+	// validate bidid
+	bidid, err := validateRequiredParameter(r, BID_ID_PARAMETER)
+
+	if err != nil {
+		return er, err
+	}
+	er.Bidid = bidid
+
+	// validate timestamp (optional)
+	err = validateTimestamp(er, r)
+
+	if err != nil {
+		return er, err
+	}
+
+	// validate format (optional)
+	err = validateFormat(er, r)
+
+	if err != nil {
+		return er, err
+	}
+
+	// validate analytics (optional)
+	err = validateAnalytics(er, r)
+
+	if err != nil {
+		return er, err
+	}
+
+	// Bidder
+	bidder := r.FormValue(BIDDER_PARAMETER)
+
+	if bidder != "" {
+		er.Bidder = bidder
+	}
+
+	return er, nil
+}
+
+func optionalParameters(request *EventRequest) string {
+	r := ""
```

Comment:
Please consider using `url.Values` from the base library instead of string concatenation. This will also take care of query string encoding.